### PR TITLE
Initial dockerfile changes for v4

### PR DESCRIPTION
**Pull request by uruwhy**
## Description
Add changes to dockerfile and docker compose file to download golang dependencies, pre-clone the atomic repo for offline usage, expose additional ports for other C2 channels

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Ran docker build (`docker build --build-arg WIN_BUILD=true . -t caldera:server`) to build an image and then ran it in a container.

Ran docker-compose up and ran operations with agents,

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by cyber-arsenull** [comment]
I tried to run `docker build --build-arg WIN_BUILD=true . -t caldera:server` with this branch, but got:

```
(caldera) PS H:\Dev\caldera> docker build --build-arg WIN_BUILD=true . -t caldera:server
[+] Building 518.2s (15/24)
 => [internal] load build definition from Dockerfile                                                                                                                                          0.1s 
 => => transferring dockerfile: 2.06kB                                                                                                                                                        0.0s 
 => [internal] load .dockerignore                                                                                                                                                             0.0s 
 => => transferring context: 35B                                                                                                                                                              0.0s 
 => [internal] load metadata for docker.io/library/ubuntu:focal                                                                                                                               1.4s 
 => [internal] load build context                                                                                                                                                             2.7s 
 => => transferring context: 3.11MB                                                                                                                                                           2.5s 
 => [ 1/20] FROM docker.io/library/ubuntu:focal@sha256:626ffe58f6e7566e00254b638eb7e0f3b11d4da9675088f4781a50ae288f3322                                                                       0.0s 
 => CACHED [ 2/20] RUN ln -snf /usr/share/zoneinfo/UTC /etc/localtime &&     echo UTC > /etc/timezone                                                                                         0.0s 
 => CACHED [ 3/20] WORKDIR /usr/src/app                                                                                                                                                       0.0s 
 => CACHED [ 4/20] RUN apt-get update &&     apt-get -y install python3 python3-pip golang git                                                                                                0.0s 
 => [ 5/20] RUN if [ "true" = "true" ] ; then apt-get -y install mingw-w64; fi                                                                                                              419.1s 
 => [ 6/20] ADD requirements.txt .                                                                                                                                                            0.3s 
 => [ 7/20] RUN pip3 install --no-cache-dir -r requirements.txt                                                                                                                              61.4s 
 => [ 8/20] ADD . .                                                                                                                                                                           5.8s 
 => [ 9/20] RUN go get github.com/grandcat/zeroconf github.com/google/go-github/github github.com/grandcat/zeroconf github.com/miekg/dns golang.org/x/oauth2 gopkg.in/natefinch/npipe.v2     29.4s 
 => [10/20] WORKDIR /usr/src/app/plugins/sandcat                                                                                                                                              0.0s 
 => ERROR [11/20] RUN ./update-agents.sh                                                                                                                                                      0.5s 
------
 > [11/20] RUN ./update-agents.sh:
#15 0.499 /bin/sh: 1: ./update-agents.sh: not found
------
executor failed running [/bin/sh -c ./update-agents.sh]: exit code: 127
```

**Comment by uruwhy** [comment]
@cyber-arsenull 

That's strange... at first I thought maybe the folders didn't get copied recursively, but it looks like it ran WORKDIR correctly. Did you try this with a fresh caldera install?

**Review by elegantmoose — APPROVED** [review]
LGTM

**Review by alexanderkent — APPROVED** [review]

File: Dockerfile [hunk 1]
```diff
@@ -13,15 +13,60 @@ RUN apt-get update && \
...
 
+# Install pip requirements
 ADD requirements.txt .
 
 RUN pip3 install --no-cache-dir -r requirements.txt
 
 ADD . .
 
+# Download golang dependencies
+RUN go get github.com/grandcat/zeroconf github.com/google/go-github/github github.com/grandcat/zeroconf github.com/miekg/dns golang.org/x/oauth2 gopkg.in/natefinch/npipe.v2
```
[the remark this answers is not in the document]
Comment (uruwhy) [reply Dockerfile:24]: good call, I can do that

File: Dockerfile [hunk 2]
```diff
@@ -13,15 +13,60 @@ RUN apt-get update && \
...
+
+# Update default sandcat agent binaries
+WORKDIR /usr/src/app/plugins/sandcat
+
+RUN ./update-agents.sh
```
[the remark this answers is not in the document]
Comment (uruwhy) [reply Dockerfile:29]: we run that shell script twice - one for a standard agent build (no extensions), and one for an agent with all extensions

**Review by alexanderkent — APPROVED** [review]
LGTM

**Review by elegantmoose — APPROVED** [review]

**Pull request merged**